src/Traits/Concerns/CalculatesCosts.php

```diff
@@ -0,0 +1,62 @@
+<?php
+
+declare(strict_types=1);
+
+namespace MichaelRubel\Couponables\Traits\Concerns;
+
+use MichaelRubel\Couponables\Exceptions\InvalidCouponTypeException;
+
+trait CalculatesCosts
+{
+    /**
+     * Calculate the output value based on the coupon type.
+     *
+     * @param  float  $value
+     *
+     * @return float
+     * @throws InvalidCouponTypeException
+     */
+    public function calc(float $value): float
+    {
+        return match ($this->{static::$bindable->getTypeColumn()}) {
+            static::TYPE_SUBTRACTION => $this->subtract($value),
+            static::TYPE_PERCENTAGE  => $this->percentage($value),
+            static::TYPE_FIXED       => $this->fixedPrice(),
+            default => throw new InvalidCouponTypeException,
+        };
+    }
+
+    /**
+     * Apply the "Subtraction" calculation strategy.
+     *
+     * @param  float  $cost
+     *
+     * @return float
+     */
+    private function subtract(float $cost): float
+    {
+        return $cost - $this->{static::$bindable->getValueColumn()};
+    }
+
+    /**
+     * Apply the "Percentage" calculation strategy.
+     *
+     * @param  float  $value
+     *
+     * @return float
+     */
+    private function percentage(float $value): float
+    {
+        return ($this->{static::$bindable->getValueColumn()} / 100) * $value;
```

Comment:
```suggestion
        return (1.0 - ($this->{static::$bindable->getValueColumn()} / 100)) * $value;
```
Usually coupons are for discount so 20% means 20% less -> 80% (1 - 0.2) of original value.
So in coupons definition table you should store discount value in %% and then use calculation as above I suppose

Reply:
More over I'm not familiar with PHP but I do not see security check in case the calculated value will be less than zero, nor here nor in `fixed` or `subtraction` methods